From 2208ab5b7d529474913dadfafdf656e29e8e1f64 Mon Sep 17 00:00:00 2001
From: Patrick Gaskin <patrick@pgaskin.net>
Date: Sat, 14 Dec 2024 14:38:51 -0500
Subject: [PATCH 6/7] Replace floating toolbar with invisible touch regions

The floating toolbar, is big, ugly, finicky to use, and generally
annoying.

Replacing it with invisible touch regions at the top and bottom of the
screen (where you wouldn't usually be dragging the mouse anyways) is
much more intuitive and unobtrusive.

The top-left corner is disconnect, top is connection info, bottom-right
is mouse buttons (as a bonus, this is the same place where the close
button for it is, so the region essentially acts as a toggle), and
bottom is keyboard.

Updating:
- Ensure the IDs in v0 match public.xml.
- Test the regions to ensure everything still works (extension keyboard
  is cleared, action is executed).
- Ensure that the real toolbar and tutorial are invisible.

diff --git a/res/layout/activity_desktop.xml b/res/layout/activity_desktop.xml
index 8087a93..be8ad89 100644
--- a/res/layout/activity_desktop.xml
+++ b/res/layout/activity_desktop.xml
@@ -5,6 +5,17 @@
         <com.realvnc.viewer.android.ui.input.CapturingEditText android:id="@id/hidden_input_view" android:focusable="true" android:layout_width="1.0dip" android:layout_height="1.0dip" android:text="" android:singleLine="false" android:imeOptions="actionNone|flagNoExtractUi" />
         <com.realvnc.viewer.android.ui.input.InterceptingRelativeLayout android:orientation="vertical" android:id="@id/intercepting_relative_layout" android:layout_width="fill_parent" android:layout_height="fill_parent" android:splitMotionEvents="true">
             <com.realvnc.viewer.android.ui.scroll.DesktopView android:orientation="vertical" android:id="@id/desktop_view" android:background="@color/background_grey" android:focusable="false" android:visibility="invisible" android:layout_width="fill_parent" android:layout_height="fill_parent" />
+            <LinearLayout android:orientation="vertical" android:layout_width="fill_parent" android:layout_height="fill_parent">
+                <LinearLayout android:orientation="horizontal" android:layout_width="fill_parent" android:layout_height="0dip" android:layout_weight="2">
+                    <View android:id="@id/menu_disconnect" android:layout_height="fill_parent" android:layout_weight="1" android:layout_width="0dp" android:background="#00000000" />
+                    <View android:id="@id/menu_information" android:layout_height="fill_parent" android:layout_weight="4" android:layout_width="0dp" android:background="#00000000" />
+                </LinearLayout>
+                <View android:layout_width="fill_parent" android:layout_height="0dp" android:layout_weight="17" />
+                <LinearLayout android:orientation="horizontal" android:layout_width="fill_parent" android:layout_height="0dip" android:layout_weight="3">
+                    <View android:id="@id/menu_keyboard" android:layout_height="fill_parent" android:layout_weight="4" android:layout_width="0dp" android:background="#00000000" />
+                    <View android:id="@id/menu_mouse" android:layout_height="fill_parent" android:layout_weight="1" android:layout_width="0dp" android:background="#00000000" />
+                </LinearLayout>
+            </LinearLayout>
             <com.realvnc.viewer.android.ui.CursorView android:id="@id/cursor_view" android:visibility="gone" android:layout_width="fill_parent" android:layout_height="fill_parent" />
             <com.realvnc.viewer.android.ui.ZoomView android:id="@id/zoom_view" android:visibility="gone" android:layout_width="wrap_content" android:layout_height="wrap_content" android:layout_alignParentBottom="true" android:layout_centerHorizontal="true" />
             <LinearLayout android:orientation="vertical" android:id="@id/overlay_parent" android:layout_width="fill_parent" android:layout_height="wrap_content" android:layout_alignParentBottom="true">
diff --git a/smali/com/realvnc/viewer/android/app/DesktopActivity.smali b/smali/com/realvnc/viewer/android/app/DesktopActivity.smali
index 792209f..2ee6c39 100644
--- a/smali/com/realvnc/viewer/android/app/DesktopActivity.smali
+++ b/smali/com/realvnc/viewer/android/app/DesktopActivity.smali
@@ -4117,6 +4117,53 @@
 
     invoke-virtual {v0, v1}, Landroid/view/View;->setOnClickListener(Landroid/view/View$OnClickListener;)V
 
+    invoke-direct {v8}, Lcom/realvnc/viewer/android/app/DesktopActivity;->useInvisibleMenu()V
+
+    return-void
+.end method
+
+.method private final useInvisibleMenu()V
+    .locals 2
+
+    new-instance v1, Lcom/realvnc/viewer/android/app/FakeMenuTrampoline;
+
+    const v0, 0x7f090167  # @id/fullscreen_toolbar_menu
+    invoke-virtual {p0, v0}, Landroid/app/Activity;->findViewById(I)Landroid/view/View;
+    move-result-object v0
+    invoke-direct {v1, v0}, Lcom/realvnc/viewer/android/app/FakeMenuTrampoline;-><init>(Landroid/view/Menu;)V
+
+    const v0, 0x7f09021a  # @id/menu_keyboard
+    invoke-virtual {p0, v0}, Landroid/app/Activity;->findViewById(I)Landroid/view/View;
+    move-result-object v0
+    invoke-virtual {v0, v1}, Landroid/view/View;->setOnClickListener(Landroid/view/View$OnClickListener;)V
+
+    const v0, 0x7f09021b  # @id/menu_mouse
+    invoke-virtual {p0, v0}, Landroid/app/Activity;->findViewById(I)Landroid/view/View;
+    move-result-object v0
+    invoke-virtual {v0, v1}, Landroid/view/View;->setOnClickListener(Landroid/view/View$OnClickListener;)V
+
+    const v0, 0x7f090219  # @id/menu_information
+    invoke-virtual {p0, v0}, Landroid/app/Activity;->findViewById(I)Landroid/view/View;
+    move-result-object v0
+    invoke-virtual {v0, v1}, Landroid/view/View;->setOnClickListener(Landroid/view/View$OnClickListener;)V
+
+    const v0, 0x7f090211  # @id/menu_disconnect
+    invoke-virtual {p0, v0}, Landroid/app/Activity;->findViewById(I)Landroid/view/View;
+    move-result-object v0
+    invoke-virtual {v0, v1}, Landroid/view/View;->setOnClickListener(Landroid/view/View$OnClickListener;)V
+
+    const v1, 8  # GONE
+
+    const v0, 0x7f09013c  # @id/fabTlbrGroup (this is the parent of the entire toolbar so we don't need to modify the rest of the code)
+    invoke-virtual {p0, v0}, Landroid/app/Activity;->findViewById(I)Landroid/view/View;
+    move-result-object v0
+    invoke-virtual {v0, v1}, Landroid/view/View;->setVisibility(I)V
+
+    const v0, 0x7f09013b  # @id/fabCoachTutorial (this is the parent of the entire tutorial, ditto)
+    invoke-virtual {p0, v0}, Landroid/app/Activity;->findViewById(I)Landroid/view/View;
+    move-result-object v0
+    invoke-virtual {v0, v1}, Landroid/view/View;->setVisibility(I)V
+
     return-void
 .end method
 
diff --git a/smali/com/realvnc/viewer/android/app/FakeMenuTrampoline.smali b/smali/com/realvnc/viewer/android/app/FakeMenuTrampoline.smali
new file mode 100644
index 0000000..d12afdc
--- /dev/null
+++ b/smali/com/realvnc/viewer/android/app/FakeMenuTrampoline.smali
@@ -0,0 +1,23 @@
+.class final Lcom/realvnc/viewer/android/app/FakeMenuTrampoline;
+.super Ljava/lang/Object;
+
+.implements Landroid/view/View$OnClickListener;
+
+.field final menu:Landroid/view/Menu;
+
+.method public constructor <init>(Landroid/view/Menu;)V
+    .locals 0
+    iput-object p1, p0, Lcom/realvnc/viewer/android/app/FakeMenuTrampoline;->menu:Landroid/view/Menu;
+    invoke-direct {p0}, Ljava/lang/Object;-><init>()V
+    return-void
+.end method
+
+.method public final onClick(Landroid/view/View;)V
+    .locals 1
+    invoke-virtual {p1}, Landroid/view/View;->getId()I
+    move-result p1
+    iget-object p0, p0, Lcom/realvnc/viewer/android/app/FakeMenuTrampoline;->menu:Landroid/view/Menu;
+    const v0, 0
+    invoke-interface {p0, p1, v0}, Landroid/view/Menu;->performIdentifierAction(II)Z
+    return-void
+.end method
